File: android/app/src/main/java/com/openledger/android/ui/ai/AiQueryViewModel.kt
package com.openledger.android.ui.ai

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.openledger.android.data.remote.dto.AiQueryResponse
import com.openledger.android.data.repository.LedgerRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class AiQueryState(
    val query: String = "",
    val isLoading: Boolean = false,
    val error: String? = null,
    val response: AiQueryResponse? = null,
)

@HiltViewModel
class AiQueryViewModel @Inject constructor(
    private val repo: LedgerRepository,
) : ViewModel() {
    private val _state = MutableStateFlow(AiQueryState())
    val state: StateFlow<AiQueryState> = _state.asStateFlow()

    fun onQueryChanged(v: String) {
        _state.value = _state.value.copy(query = v, error = null)
    }

    fun submit() {
        val q = _state.value.query.trim()
        if (q.isBlank()) return
        _state.value = _state.value.copy(isLoading = true, error = null, response = null)
        viewModelScope.launch {
            repo.aiQuery(q)
                .onSuccess { _state.value = _state.value.copy(isLoading = false, response = it) }
                .onFailure { _state.value = _state.value.copy(isLoading = false, error = it.message) }
        }
    }
}

File: ios/OpenLedger/OpenLedger/ViewModels/AiQueryViewModel.swift
import Foundation

@Observable
final class AiQueryViewModel {
    var query = ""
    var isLoading = false
    var error: String?
    var response: AiQueryResponse?

    private let repo: LedgerRepository

    init(repo: LedgerRepository) {
        self.repo = repo
    }

    func submit() async {
        let q = query.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !q.isEmpty else { return }
        isLoading = true
        error = nil
        response = nil
        do {
            response = try await repo.aiQuery(q)
            isLoading = false
        } catch {
            self.error = error.localizedDescription
            isLoading = false
        }
    }
}
